File: src/com/java/test/Singleton.java
package com.java.test;

public class Singleton {


    private Singleton() {
    }

    // ================================================== //
    private static Singleton HUNGRY = new Singleton();

    public static Singleton hungry() {
        return HUNGRY;
    }

    // ================================================== //
    private volatile static Singleton LAZY;

    public static Singleton lazy() {
        if (LAZY == null) {
            synchronized (Singleton.class) {
                if (LAZY == null) {
                    LAZY = new Singleton();
                }
            }
        }
        return LAZY;
    }

    // ================================================== //

    public static class Holder {
        private static Singleton HOLDER = new Singleton();
    }

    public Singleton getHolder() {
        return Holder.HOLDER;
    }
    // ================================================== //

    public enum SingletonEnum {
        INSTANCE;
    }
}

File: src/com/kotlin/test/Singleton.kt
package com.kotlin.test

/**
 * 饿汉式单例
 */
object UrgentSingleton {

    fun foo() = "UrgentSingleton"
}

/**
 * 懒汉式单例
 */
class LazySingleton private constructor() {

    fun foo() = "LazySingleton"

    companion object {
        val instance: LazySingleton by lazy { LazySingleton() }
    }


}

fun main(args: Array<String>) {
    UrgentSingleton.foo()
    LazySingleton.instance.foo()
}

